A. Commit message:
Fix issue of listen before chmod on Unix sockets (CVE-2023-45145)

B. Change summary:
The patch ensures that the permissions for Unix sockets are set up prior to calling `listen()`. This change prevents a race condition where another process could exploit the socket's temporary incorrect permissions.

C. Compatibility Risk:
Medium

D. Fixed Code:
diff --git a/src/anet.c b/src/anet.c
index 4ea201df57f..f2e8a6d9c0a 100644
--- a/src/anet.c
+++ b/src/anet.c
@@ -407,13 +407,16 @@ int anetUnixGenericConnect(char *err, const char *path, int flags)
     return s;
 }
 
-static int anetListen(char *err, int s, struct sockaddr *sa, socklen_t len, int backlog) {
+static int anetListen(char *err, int s, struct sockaddr *sa, socklen_t len, int backlog, mode_t perm) {
     if (bind(s,sa,len) == -1) {
         anetSetError(err, "bind: %s", strerror(errno));
         close(s);
         return ANET_ERR;
     }
 
+    if (sa->sa_family == AF_LOCAL && perm)
+        chmod(((struct sockaddr_un *) sa)->sun_path, perm);
+
     if (listen(s, backlog) == -1) {
         anetSetError(err, "listen: %s", strerror(errno));
         close(s);